tmt/beakerlib.py

```diff
@@ -100,20 +120,21 @@ def __init__(self, identifier, parent=None):
                 raise tmt.utils.SpecificationError(
                     f"Library name '{self.name}' does not start with a '/'.")
             # Use provided repository nick name or parse it from the url/path
-            self.repo = identifier.get('nick')
-            if not self.repo and self.url:
-                try:
-                    self.repo = re.search(
-                        r'/([^/]+?)(/|\.git)?$', self.url).group(1)
-                except AttributeError:
+            repo = identifier.get('nick')
+            if not repo and self.url:
+                repo_search = re.search(r'/([^/]+?)(/|\.git)?$', self.url)
+                if not repo_search:
                     raise tmt.utils.GeneralError(
                         f"Unable to parse repository name from '{self.url}'.")
-            if not self.repo and self.path:
+                repo = repo_search.group(1)
+            if not repo and self.path:
                 try:
-                    self.repo = os.path.basename(self.path)
+                    repo = os.path.basename(self.path)
                 except TypeError:
                     raise tmt.utils.GeneralError(
                         f"Unable to parse repository name from '{self.path}'.")
+            assert repo
```

Comment:
This is suspicious. At least one of `self.url` and `self.path` shall be set, according to line 106. Ok, let's say it's `self.url` and `self.path` is `None`. Then line 123, `repo` is read from `identifier`, let's say it's empty or `None`-ish - apparently, that can happen, because on line 124 there is a check for `not repo`. `self.url` is set, so we dive under the `if`, and let's say `repo_search` is valid, not `None`, but `repo_search.group(1)` is an empty string - can this happen? If so, then `not repo and self.path` condition on line 130 will not trigger, because `self.path` is `None` as we already know, and `repo` will not pass the `assert repo` check you added.

In other words, this `assert` suggests the code flow has some shady area which prevents mypy from deducing `repo` is non-empty value by now. If `repo == ""` is allowed (can it happen on line 129?), then this assert shall be `assert repo is not None`. Or, if the block on like 124 fails to produce non-empty `repo`, with `self.path is None`, the check on like 130 needs an update.

Reply:
I see. Then maybe the condition on line 130 should be `elif`: if `repo` is either already set or comes set from the block above, then `if not repo` can never trigger. Or maybe even refactored, to help mypy (and, obviously, me :), to something like:

```python
if not repo:
  if self.url:
    ...
  else:
   ...

# assert repo - maybe this one won't be needed
self.repo = repo
```

Right, so with this it makes sense, since mypy cannot know the regex won't let this happen. With the refactoring, maybe the assert won't be needed at all, but it's been certainly a bit complicated to me.